pkg/controller/configmapcainjector/configmap_injecting_controller.go

```diff
@@ -0,0 +1,178 @@
+package configmapcainjector
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/api/core/v1"
+	kapierrors "k8s.io/apimachinery/pkg/api/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	informers "k8s.io/client-go/informers/core/v1"
+	kcoreclient "k8s.io/client-go/kubernetes/typed/core/v1"
+	listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const (
+	InjectCABundleAnnotation = "service.alpha.openshift.io/inject-cabundle"
+	InjectionDataKey         = "service-ca.crt"
+)
+
+// ConfigMapCABundleInjectionController is responsible for injecting a CA bundle into configMaps annotated with
+// "service.alpha.openshift.io/inject-cabundle"
+type ConfigMapCABundleInjectionController struct {
+	configMapClient kcoreclient.ConfigMapsGetter
+
+	// configMaps that need to be checked
+	queue workqueue.RateLimitingInterface
+
+	configMapLister    listers.ConfigMapLister
+	configMapHasSynced cache.InformerSynced
+
+	ca string
+
+	// syncHandler does the work. It's factored out for unit testing
+	syncHandler func(serviceKey string) error
+}
+
+// NewConfigMapCABundleInjectionController creates a new ServiceServingCertUpdateController.
+// TODO this should accept a shared informer
+func NewConfigMapCABundleInjectionController(configMaps informers.ConfigMapInformer, configMapsClient kcoreclient.ConfigMapsGetter, ca []byte, resyncInterval time.Duration) *ConfigMapCABundleInjectionController {
+	ic := &ConfigMapCABundleInjectionController{
+		queue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		ca:    string(ca),
+	}
+
+	ic.configMapLister = configMaps.Lister()
+	configMaps.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    ic.addConfigMap,
+			UpdateFunc: ic.updateConfigMap,
+		},
+		resyncInterval,
+	)
+	ic.configMapHasSynced = configMaps.Informer().GetController().HasSynced
+	ic.configMapClient = configMapsClient
+	ic.syncHandler = ic.syncConfigMap
+
+	return ic
+}
+
+// Run begins watching and syncing.
+func (ic *ConfigMapCABundleInjectionController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer ic.queue.ShutDown()
+
+	// Wait for the stores to fill
+	if !cache.WaitForCacheSync(stopCh, ic.configMapHasSynced) {
+		return
+	}
+
+	glog.V(1).Infof("Starting workers")
+	for i := 0; i < workers; i++ {
+		go wait.Until(ic.runWorker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.V(1).Infof("Shutting down")
+}
+
+func (ic *ConfigMapCABundleInjectionController) enqueueConfigMap(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+
+	ic.queue.Add(key)
+}
+
+func (ic *ConfigMapCABundleInjectionController) addConfigMap(obj interface{}) {
+	cm, ok := obj.(*v1.ConfigMap)
+	if !ok {
```

Comment:
Thoughts on adding debug lines for these short circuit cases?

Reply:
yep, I'll add some.